### PR TITLE
Add manifestWorkReplicaSets controller

### PR DESCRIPTION
**Pull request by serngawy**

### CODE REVIEW

File: manifests/cluster-manager/hub/cluster-manager-registration-clusterrole.yaml [hunk 1]
```diff
@@ -42,6 +42,10 @@ rules:
 - apiGroups: ["work.open-cluster-management.io"]
   resources: ["manifestworks"]
   verbs: ["get", "list", "watch"]
+# Allow hub to monitor manifestworkreplicasets
+- apiGroups: ["work.open-cluster-management.io"]
```
Comment (qiujian16) [inline manifests/cluster-manager/hub/cluster-manager-registration-clusterrole.yaml:46]: you also need manifestworkreplicasets/status, and manifestworkreplicasets/finalize permission I think

**Review by qiujian16 — COMMENTED** [review]
I think we need a flag on clustermanager api to enable it, and we should disable it by default in the alpha stage. like this one https://github.com/open-cluster-management-io/api/pull/211

File: manifests/cluster-manager/hub/cluster-manager-registration-clusterrole.yaml (hunk 1 above)
Comment (serngawy) [reply manifests/cluster-manager/hub/cluster-manager-registration-clusterrole.yaml:46]: done

**Comment by serngawy** [comment]
> I think we need a flag on clustermanager api to enable it, and we should disable it by default in the alpha stage. like this one [open-cluster-management-io/api#211](https://github.com/open-cluster-management-io/api/pull/211)

okay, but the manifestWorkReolicaSet it is already part of the work repo I believe its belong to workConfiguration . is it will be as [FeatureGate](https://github.com/open-cluster-management-io/api/blob/main/operator/v1/types.go#L88) ?  if yes, what name it should has ?

**Comment by qiujian16** [comment]
maybe a featureGate EnableReplicaSetManager

**Comment by serngawy** [comment]
> maybe a featureGate EnableReplicaSetManager

fix it, it has nothing to do with the featureGate , the integration test failed due to number of CRDs created.

**Review by qiujian16 — COMMENTED** [review]
you would also need to start the manager on the hub side. we currently only start webhook.

**Comment by serngawy** [comment]
> you would also need to start the manager on the hub side. we currently only start webhook.

Would you advice Jian I'm kind of lost between the work repo and this one.
 I added the manager-deployment and the manifestWorkReplicaSet-webhookvalidation. Do we need to add service and endpoint as well ?

**Comment by qiujian16** [comment]
no, we do not need service

File: manifests/cluster-manager/management/cluster-manager-manifestworkreplicaset-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        args:
+          - "/work"
+          - "manager"
+        securityContext:
+          allowPrivilegeEscalation: false
+          capabilities:
+            drop:
+              - ALL
+          privileged: false
+          runAsNonRoot: true
```
Comment (qiujian16) [inline manifests/cluster-manager/management/cluster-manager-manifestworkreplicaset-deployment.yaml:58]: also need livenessprobe/readinessprobe and resource request. We also need image part as a template variable.

File: manifests/cluster-manager/hub/cluster-manager-manifestworkreplicaset-webhook-validatingconfiguration-v1alpha1.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: manifestworkreplicasetv1alpha1validators.admission.work.open-cluster-management.io
+webhooks:
+- name: manifestworkreplicasetv1alpha1validators.admission.work.open-cluster-management.io
+  failurePolicy: Fail
+  clientConfig:
+    service:
+      namespace: {{ .ClusterManagerNamespace }}
+      name: cluster-manager-registration-webhook
```
Comment (qiujian16) [inline manifests/cluster-manager/hub/cluster-manager-manifestworkreplicaset-webhook-validatingconfiguration-v1alpha1.yaml]: I think the service name should be cluster-manager-work-webhook

File: manifests/cluster-manager/hub/cluster-manager-manifestworkreplicaset-webhook-validatingconfiguration.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+apiVersion: admissionregistration.k8s.io/v1
```
Comment (qiujian16) [inline manifests/cluster-manager/hub/cluster-manager-manifestworkreplicaset-webhook-validatingconfiguration.yaml:1]: why we need two same configurations?

Comment (serngawy) [reply manifests/cluster-manager/hub/cluster-manager-manifestworkreplicaset-webhook-validatingconfiguration.yaml:1]: ops , forget to delete it

File: pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go [hunk 5]
```diff
@@ -29,6 +29,7 @@ var (
 		"cluster-manager/management/cluster-manager-registration-webhook-deployment.yaml",
 		"cluster-manager/management/cluster-manager-work-webhook-deployment.yaml",
 		"cluster-manager/management/cluster-manager-placement-deployment.yaml",
+		"cluster-manager/management/cluster-manager-manifestworkreplicaset-deployment.yaml",
```
Comment (zhiweiyin318) [inline pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go:32]: should we check the featureGate EnableReplicaSetManager to determine if install the manifestworkreplicaset manifests like  https://github.com/open-cluster-management-io/registration-operator/blob/547fc1482e257e7eb8abfcd13c456513837d5574/pkg/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_runtime_reconcile.go#L106 ?

**Comment by qiujian16** [comment]
@zhiweiyin318 let's merge it at first, We could handle the comment as the followup

/approve
/lgtm

**Pull request merged**